### PR TITLE
Add HostAliases to PodSpec to support adding entires to a Pod's hosts file

### PR DESCRIPTION
**Pull request by evie404**
**What this PR does / why we need it**:
Adds a new field to PodSpec `HostAliases ` to support adding entries to a Pod's hosts file. A PR to incorporate this logic into [`ensureHostsFile`](https://github.com/kubernetes/kubernetes/blob/2899f47bc86695ac8093c66cef2cc02d55c6a23d/pkg/kubelet/kubelet_pods.go#L208-L227) in kubelet will be next in order to isolate the discussion on the API.

**Which issue this PR fixes**:
A step into fixing #43632

**Special notes for your reviewer**:
- The API was suggested in https://github.com/kubernetes/kubernetes/issues/43632#issuecomment-293471293
- Concerns with compatible with `hostNetwork: true` Pods are addressed with an validation. https://github.com/kubernetes/kubernetes/issues/43632#issuecomment-293750410

**Release note**:
```release-note
A new field `hostAliases` has been added to `pod.spec` to support adding entries to a Pod's /etc/hosts file.
```

**Testing done**:
- Unit tests pass for validations
- Ran all automated updates, and compared against another PR which added a field in PodSpec to make sure the PR contains all the updated generated code

### CODE REVIEW

**Comment by rickypai** [comment]
Btw, this should probably inherit the same labels as #43632:
- area/kubelet
- kind/enhancement
- sig/node

**Comment by rickypai** [comment]
@thockin hi tim, is it possible for you to help me look at this change? it implements the proposed type for host mapping in `/etc/hosts`: https://github.com/kubernetes/kubernetes/issues/43632#issuecomment-293471293

**Comment by dims** [comment]
@k8s-bot ok to test

**Comment by rickypai** [comment]
seems like I forgot to generate federation docs and also one of the fields do not conform to convention. updating those.

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -1909,6 +1909,16 @@ func validateHostNetwork(hostNetwork bool, containers []api.Container, fldPath *
 	return allErrors
 }
 
+func validateHostNetworkNoHostMappings(hostNetwork bool, hostMappings []api.HostMapping, fldPath *field.Path) field.ErrorList {
+	allErrors := field.ErrorList{}
+	if hostNetwork {
+		if len(hostMappings) > 0 {
+			allErrors = append(allErrors, field.Invalid(fldPath, hostMappings, "no `hostMappings` allowed when `hostNetwork` is true"))
```
Comment (thockin) [inline pkg/api/validation/validation.go]: field.Forbidden(..."may not be set when `hostNetwork` is true")

**Review by thockin — COMMENTED** [review]
I'll approve and let the reviewer's LGTM

**Comment by thockin** [comment]
@yujuhong I'm open to better names, and if you want to discuss this more, that's fair - just hold off on the LGTM.

File: pkg/api/validation/validation.go (hunk 1 above)
Comment (rickypai) [inline pkg/api/validation/validation.go]: fixed

**Comment by rickypai** [comment]
Updating the name to `HostAliases` per discussion in https://github.com/kubernetes/kubernetes/issues/43632#issuecomment-297875527

**Comment by rickypai** [comment]
@k8s-bot unit test this

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -1909,6 +1909,16 @@ func validateHostNetwork(hostNetwork bool, containers []api.Container, fldPath *
 	return allErrors
 }
 
+func validateHostNetworkNoHostAliases(hostNetwork bool, hostAliases []api.HostAlias, fldPath *field.Path) field.ErrorList {
+	allErrors := field.ErrorList{}
+	if hostNetwork {
+		if len(hostAliases) > 0 {
+			allErrors = append(allErrors, field.Forbidden(fldPath, "`hostAliases` may not be set when `hostNetwork` is true"))
```
Comment (thockin) [inline pkg/api/validation/validation.go]: don't lead with the field name - it will already be present in the structured error

**Review by thockin — COMMENTED** [review]
sorry to nitpick.

**Comment by thockin** [comment]
Just one nit and I am LGTM.  Thanks!

**Comment by thockin** [comment]
This would make a great Task doc - @chenopis

File: pkg/api/validation/validation.go (hunk 2 above)
Comment (rickypai) [inline pkg/api/validation/validation.go]: fixed

**Comment by rickypai** [comment]
@k8s-bot unit test this

**Comment by rickypai** [comment]
@k8s-bot unit test this

**Comment by rickypai** [comment]
@k8s-bot unit test this

**Comment by rickypai** [comment]
@k8s-bot unit test this

**Comment by rickypai** [comment]
@k8s-bot unit test this

**Comment by rickypai** [comment]
Keep getting unrelated test failures. Gonna rebase master and see if it helps.

**Comment by rickypai** [comment]
@k8s-bot unit test this

**Comment by rickypai** [comment]
@thockin i've addressed the issues and all tests pass. if you can take a look at this again that'd be great!

**Comment by thockin** [comment]
Uggh, another rebase.  Sorry

**Comment by rickypai** [comment]
np. thanks for the review!

**Comment by rickypai** [comment]
@k8s-bot kubemark e2e test this

**Comment by rickypai** [comment]
Thanks a lot for the fast review! There's discussion in #43632 about
whether e2e is actually needed. I can write an e2e or integration test when
the dust settles on which one is preferred

Also, Do you think this is a valid candidate for backport to 1.6? It
addresses a lot of user issues around /etc/hosts management.

Thanks a lot again!

**Comment by rickypai** [comment]
Opps the above comment was meant for a different PR.

**Comment by WanLinghao** [comment]
@rickypai  thanks for your well done job!
And  may I ask you that file like  pkg/api/v1/types.generated.go was auto-generated by what tool?
thanks a lot!

**Comment by rickypai** [comment]
@WanLinghao https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md

**Comment by WanLinghao** [comment]
@rickypai  Thanks!

**Pull request merged**